### ISSUE TITLE
Invalid memory read

### ISSUE THREAD

**Issue by dwilliamson** [sp4cerat/Fast-Quadric-Mesh-Simplification#2]
There's a `push_back` on [line #543](https://github.com/sp4cerat/Fast-Quadric-Mesh-Simplification/blob/master/src.cmd/Simplify.h#L543) that uses a reference from an object in the same vector on [line #528](https://github.com/sp4cerat/Fast-Quadric-Mesh-Simplification/blob/master/src.cmd/Simplify.h#L528). If the vector needs to reallocate the memory of `refs` to be big enough for the new object, the reference to the source `Ref` object becomes invalid.

Fix is to simply remove the reference and copy the `Ref` object.

**Comment by neurolabusc** [comment]
Am I correct that your proposed fix is to change the line
 Ref &r=refs[v.tstart+k];
to read
 Ref r=refs[v.tstart+k];
So that 'r' is a struct rather than a pointer. The issue with the pointer is that push_back might reallocate the array so a pointer is no longer valid.

**Comment by dwilliamson** [comment]
Are you misunderstanding the original source or why this is a bug?

**Comment by neurolabusc** [comment]
Neither, just suggesting how your fix could be implemented (e.g. removing the &). I think this is the minimal change and keeps the code the most readable. However, I suspect there is a very small (probably negligible) cost to copying to a new structure, so the other alternative is to use the pointer initially and only copy the structure after the two continue statements.

By the way, I did notice, fix and comment this issue when I ported this code to Pascal. However, I did not report an issue for the C code because I never encountered a problem with the compiled C code, and assumed that the C compiler was doing some very clever footwork for the push_back command. In hindsight, I should have reported this, as even if it has worked every time I tested it with my compiler it might fail with other datasets (depending on memory management) or other compilers.
